doc/gold.rst

```diff
@@ -0,0 +1,412 @@
+
+The binding introduced by a ``use`` statement is not exported from the unit.
+
+Units have names derived from the name of the file they are defined in, and thus
+the allowed names for units may be further restricted by what characters are
+allowed in filenames on your system. For maximum portability, stick to
+alphanumeric characters and the character ``-`` (hyphen minus) in your unit
+names.
+
+
+``constant``
+~~~~~~~~~~~~
+
+Top-level / block-level usage: ::
+
+  constant <identifier> : <type> = <known-expression>
+
+Binds a name to a value known at compile time which does not allocate memory in
+the program image. The value will be inlined at usage sites. Top-level constant
+bindings are exported from the unit as symbols, and may be referenced in other
+units.
+
+The restriction to values which do not allocate memory means that arrays and
+strings cannot be declared as constant-bindings. It is possible to declare
+pointers and slices as constants through the use of certain functions exported
+from the built-in ``mem`` unit.
+
+
+``let``
+~~~~~~~
+
+Top-level usage ::
+
+  let [mut] <identifier> : <type> = <known-expression>
+
+Binds a name to a value known at compile time. Always allocates memory in the
+program image. Top-level let-bindings are exported from the unit as symbols
+which may be referenced in other units.
+
+Block-level usage: ::
+
+  let [mut] <identifier> : <type> = <expression>
+  let stash <identifier> : <type> = <known-expression>
+
+Binds a name to a value. In the first form, memory is allocated statically (i.e.
+memory is reserved, but the value is not included in the program image), and the
+value is computed and stored when the statement is executed. In the second form,
+memory is allocated in the program image with the initial value stored.
+
+By default, let-bindings are immutable, and may not be re-bound (though they may
+be shadowed). If the ``mut`` or ``stash`` storage classes are applied, then the
+binding becomes mutable, and the value may be changed.
+
+
+``fun``
+~~~~~~~
+
+Top-level usage: ::
+
+  fun <identifier>([<identifier> : <type> [, ...]]) [-> <type-expression>]
+    [...]
+  endfun
+
+Binds a name to a function with zero or more arguments and an optional return
+type. Introduces a new scope, and statements inside are considered block-level
+statements.
+
+A function with a return type must terminate by executing a ``return``
+statement.
+
+Note that the type of the binding introduced is a function type. Function types
+may only be used to call the function or get a pointer to its address using the
+``&`` operator.
+
+
+``isr``
+~~~~~~~
+
+Top-level usage: ::
+
+  isr <identifier>
+    [...]
+  endisr 
+
+Binds a name to an interrupt service routine. Introduces a new scope, and
+statements inside are considered block-level statements.
+
+Note that the type of the binding introduced is an ISR type. ISR types may only
+be used to get a pointer to its address using the ``&`` operator.
+
+
```

Comment:
did we dump the `refun` keyword?

Reply:
No, I just missed it out. I wanted to wait until we've nailed down the calling conventions and `redo` before tackling it. Calling convention concerns will probably hobble it, and `redo` will probably take over some of it's use cases.